## Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
df.to_csv("chartiy_data.csv")
df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [2]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.

df = df.drop(columns=['EIN', 'NAME'])

df

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
0       Association       1              0                      N      5000   
1      Co-operative       1         1-9999                      N    108590   
2       Association       1              0                      N      5000   
3             Trust       1    10000-24999                      N      6692   
4             Trust       1  100000-499999                      N    142590   
...             ...     ...            ...                    ...       ...   
34294   Association       1              0                      N      5000   
34295   Association       1              0                      N      5000   
34296   Association       1              0                      N      5000   
34297   Association       1              0                      N      5000   
34298  Co-operative       1          1M-5M                      N  36500179   

       IS_SUCCESSFUL  
0                  1  
1                  1  
2                  0  
3                  1  
4                  1  
...              ...  
34294              0  
34295              0  
34296              0  
34297              1  
34298              0  

[34299 rows x 10 columns]

In [3]:
# Determine the number of unique values in each column.

unique_values = df.nunique()
print(unique_values)

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64


In [4]:
# Look at APPLICATION_TYPE value counts to identify and replace with "Other"

application_type_counts = df['APPLICATION_TYPE'].value_counts()
print(application_type_counts)

APPLICATION_TYPE
T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: count, dtype: int64


In [5]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`

cutoff_value = 500
application_types_to_replace = application_type_counts[application_type_counts < cutoff_value].index.tolist()

# Replace in dataframe
for app in application_types_to_replace:
    df['APPLICATION_TYPE'] = df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure replacement was successful
df['APPLICATION_TYPE'].value_counts()

APPLICATION_TYPE
T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: count, dtype: int64

In [6]:
# Look at CLASSIFICATION value counts to identify and replace with "Other"

classification_counts = df['CLASSIFICATION'].value_counts()
print(classification_counts)

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: count, Length: 71, dtype: int64


In [7]:
# You may find it helpful to look at CLASSIFICATION value counts >1

classification_counts_above_1 = classification_counts[classification_counts > 1]
print(classification_counts_above_1)

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: count, dtype: int64


In [8]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`

cutoff_value = 1000
classifications_to_replace = classification_counts[classification_counts < cutoff_value].index.tolist()

# Replace in dataframe
for cls in classifications_to_replace:
    df['CLASSIFICATION'] = df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure replacement was successful
df['CLASSIFICATION'].value_counts()

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: count, dtype: int64

In [9]:
# Convert categorical data to numeric with `pd.get_dummies`

df_encoded = pd.get_dummies(df)

In [10]:
# Split our preprocessed data into our features and target arrays

y = df_encoded['IS_SUCCESSFUL'].values
X = df_encoded.drop('IS_SUCCESSFUL', axis=1).values

# Split the preprocessed data into a training and testing dataset

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [11]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [12]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=80, activation='relu', input_dim=X_train_scaled.shape[1]))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=30, activation='relu'))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn.summary()

C:\Users\twpot\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense (Dense)                        │ (None, 80)                  │           3,520 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 30)                  │           2,430 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 1)                   │              31 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 5,981 (23.36 KB)

 Trainable params: 5,981 (23.36 KB)

 Non-trainable params: 0 (0.00 B)

In [13]:
# Compile the model

nn.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [14]:
# Train the model

fit_model = nn.fit(X_train_scaled, y_train, epochs=100, batch_size=32, validation_split=0.1)

Epoch 1/100
772/772 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.6987 - loss: 0.5959 - val_accuracy: 0.7354 - val_loss: 0.5446
Epoch 2/100
772/772 ━━━━━━━━━━━━━━━━━━━━ 1s 921us/step - accuracy: 0.7340 - loss: 0.5493 - val_accuracy: 0.7351 - val_loss: 0.5464
Epoch 3/100
772/772 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7285 - loss: 0.5539 - val_accuracy: 0.7362 - val_loss: 0.5475
Epoch 4/100
772/772 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7340 - loss: 0.5476 - val_accuracy: 0.7380 - val_loss: 0.5441
Epoch 5/100
772/772 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7345 - loss: 0.5487 - val_accuracy: 0.7289 - val_loss: 0.5445
Epoch 6/100
772/772 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7358 - loss: 0.5417 - val_accuracy: 0.7380 - val_loss: 0.5419
Epoch 7/100
772/772 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7323 - loss: 0.5470 - val_accuracy: 0.7358 - val_loss: 0.5428
Epoch 8/100
772/772 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7327 - loss: 0.5442 - val_ac

Epoch 51/100
772/772 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7339 - loss: 0.5429 - val_accuracy: 0.7409 - val_loss: 0.5381
Epoch 52/100
772/772 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7360 - loss: 0.5396 - val_accuracy: 0.7394 - val_loss: 0.5401
Epoch 53/100
772/772 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7357 - loss: 0.5387 - val_accuracy: 0.7409 - val_loss: 0.5383
Epoch 54/100
772/772 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7415 - loss: 0.5363 - val_accuracy: 0.7398 - val_loss: 0.5402
Epoch 55/100
772/772 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7451 - loss: 0.5295 - val_accuracy: 0.7398 - val_loss: 0.5438
Epoch 56/100
772/772 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7392 - loss: 0.5353 - val_accuracy: 0.7398 - val_loss: 0.5384
Epoch 57/100
772/772 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7419 - loss: 0.5325 - val_accuracy: 0.7402 - val_loss: 0.5429
Epoch 58/100
772/772 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7370 - loss: 0.5360 - 

In [15]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

215/215 - 0s - 814us/step - accuracy: 0.7281 - loss: 0.5683
Loss: 0.5683092474937439, Accuracy: 0.7281340956687927


In [16]:
# Export our model to HDF5 file

nn.save("AlphabetSoupCharity.h5")

# Optimization

In [17]:
# This function is responsible for creating a new sequential model with customizable hyperparameters
def build_model(hp):
    nn_model = tf.keras.models.Sequential()

    # Determine the best actitivation function
    activation = hp.Choice('activation', ['relu', 'tanh', 'sigmoid'])

    # Determine the best number of units
    layer_units = hp.Choice('first_units', [16, 32, 48, 64, 80])
    dense_layer = tf.keras.layers.Dense(units=layer_units, activation=activation, input_dim=X_train_scaled.shape[1])
    nn_model.add(dense_layer)

    # Determine the best number of hidden layers
    for i in range(hp.Int('num_layers', 1, 3)):
        layer_units = hp.Choice(f'units_{i}', [8, 12, 16, 20, 24])
        dense_layer = tf.keras.layers.Dense(units=layer_units, activation=activation)
        nn_model.add(dense_layer)

    dense_layer = tf.keras.layers.Dense(units=1, activation="sigmoid")
    nn_model.add(dense_layer)

    # Compile the model
    nn_model.compile(loss="binary_crossentropy", optimizer='adam', metrics=["accuracy"])

    return nn_model

In [18]:
# Tune with keras tuner

import keras_tuner as kt

tuner = kt.Hyperband(
    build_model,
    objective = "val_accuracy",
    max_epochs = 50,
    hyperband_iterations = 2
)

tuner.search(X_train_scaled,y_train,epochs=50,validation_data=(X_test_scaled,y_test))

best_model = tuner.get_best_hyperparameters()[0]
best_model.values

Trial 73 Complete [00h 00m 43s]
val_accuracy: 0.7294460535049438

Best val_accuracy So Far: 0.7310495376586914
Total elapsed time: 00h 10m 13s

Search: Running Trial #74

Value             |Best Value So Far |Hyperparameter
relu              |relu              |activation
48                |64                |first_units
3                 |3                 |num_layers
24                |24                |units_0
12                |24                |units_1
12                |24                |units_2
50                |50                |tuner/epochs
17                |17                |tuner/initial_epoch
2                 |3                 |tuner/bracket
2                 |3                 |tuner/round
0067              |0047              |tuner/trial_id

Epoch 18/50
858/858 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.7340 - loss: 0.5427 - val_accuracy: 0.7280 - val_loss: 0.5593
Epoch 19/50
858/858 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7335 - loss: 0.5448 - val_acc

KeyboardInterrupt: 

In [23]:
# Test optimized model accuracy to verify it exceeds 75%

best_model = tuner.get_best_models(1)[0]
model_loss, model_accuracy = best_model.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")
tuner.get_best_hyperparameters()[0].values

215/215 - 0s - 2ms/step - accuracy: 0.7310 - loss: 0.5563
Loss: 0.556309700012207, Accuracy: 0.7310495376586914


{'activation': 'relu',
 'first_units': 64,
 'num_layers': 3,
 'units_0': 24,
 'units_1': 24,
 'units_2': 24,
 'tuner/epochs': 50,
 'tuner/initial_epoch': 17,
 'tuner/bracket': 3,
 'tuner/round': 3,
 'tuner/trial_id': '0047'}

In [20]:
# Save the optimized model

best_model.save('AlphabetSoupCharity_Optimization.h5')

# Report

Purpose: The purpose of this analysis is to develop a customizable neural network model for binary classification tasks, with the aim of finding an optimal model configuration through hyperparameter tuning.

Questions: The "Starter Code" folder does not appear to contain 6 questions that I can answer

Summary: There are millions of combinations that could be tested. Out of the ones that were actually tested, the best one used a relu activation function with 64 first units, 3 layers, and 24 units for each layer. This produced an accuracy of 73.10%.

Alternatives: For this classification task, we could also use logistic regression, random forest classifier, of XGBoost classifier. If you have the computational resources, you would try all combinations and use the model with the best predictive accuracy.